---
title: "roe"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{roe}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
suppressPackageStartupMessages({
  library(roe)
})
```

# Introduction

[`Alevin-fry`](https://github.com/COMBINE-lab/alevin-fry) is a fast, accurate, and memory frugal quantification tool for preprocessing single-cell RNA-sequencing data. Detailed information can be found in the alevin-fry [manuscript](https://www.biorxiv.org/content/10.1101/2021.06.29.450377v2).

The `roe` package provides useful functions for preparing input files required by `alevin-fry`, which consists of:\
1. preparing the *splici* reference for the `USA` mode of alevin-fry, which will export a unspliced, a spliced, and an ambiguous molecule count for each gene within each cell.

# Installlation

The `roe` package can be accessed from its [github repository](https://github.com/COMBINE-lab/roe). To install the `roe` package, start R and enter

```{r install_roe, eval=FALSE}

# make sure the dependencies are installed
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

# Some functions in roe are highly relied on eisaR package, so install it to get all the dependencies.
# Install eisaR
BiocManager::install("eisaR")

# install roe from github
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")

# Install roe
devtools::install_github("COMBINE-lab/roe")
```


# Preparing a splici index for quantification with alevin-fry

The USA mode in alevin-fry requires a special index reference, which is called the *splici* reference. The *splici* reference contains the spliced transcripts plus the intronic sequences of each gene. The `make_splici_txome()` function is designed to make the *splici* reference by taking a genome FASTA file and a gene annotation GTF file as the input. Details about the *splici* can be found in Section S2 of the supplementary file of the [alevin-fry manuscript](https://www.biorxiv.org/content/10.1101/2021.06.29.450377v2). To run the function, you also need to specify the read length argument `read_length` of the experiment you are working on and the flank trimming length `flank_trim_length`. A final flank length will be computed as the difference between the read_length and flank trimming length and will be attached to the ends of each intron to absorb the intron-exon junctional reads. This function is based on the `getFeatureRanges()` function in [`eisaR`](https://github.com/fmicompbio/eisaR) package.

Following is an example of calling the `make_splici_txome()` to make the *splici* index reference. The final flank length is calculated as the difference between the read length and the flank_trim_length, i.e., $5-2=3$. This function allows you to add extra spliced and unspliced sequences to the *splici* index, which will be useful when some unannotated sequences, such as mitochondrial genes, are important for your experiment. 

```{r}
genome_path <- system.file("extdata/small_example_genome.fa", package = "roe")
gtf_path <- system.file("extdata/small_example.gtf", package = "roe")
extra_spliced = system.file("extdata/extra_spliced.txt", package = "roe")
extra_unspliced = system.file("extdata/extra_unspliced.txt", package = "roe")
output_dir = tempdir()
read_length=5
flank_trim_length = 2
filename_prefix = "transcriptome_splici"

# run the function
make_splici_txome(genome_path = genome_path,
                  gtf_path = gtf_path,
                  read_length = read_length,
                  output_dir = output_dir,
                  flank_trim_length = flank_trim_length,
                  filename_prefix = filename_prefix,
                  extra_spliced = extra_spliced,
                  extra_unspliced = extra_unspliced,
                  dedup_seqs = FALSE,
                  no_flanking_merge = FALSE
) 

# grep the output filenames
grep("transcriptome_splici_fl", dir(output_dir), value = TRUE)

```

The `make_splici_txome()` function has no returned value, but writes three files to your specified output directory `output_dir`. They are 
- A FASTA file that stores the extracted splici sequences.
- A transcript name to gene name file that stores the corresponding gene name for each transcript.
- A three columns' transcript name to gene name file that stores 

## the *splici* index

The *splici* index of a given species consists of the transcriptome of the species, i.e., the spliced transcripts, and the intronic sequences of the species. Within a gene, if the flanked intronic sequences overlap with each other, the overlapped intronic sequences will be collapsed as a single intronic sequence to make sure each base will appear only once in the intronic sequences.

### Some details about *splici*


The term *splici* is shorthand for spliced + intronic reference sequence. This reference sequence is prepared by extracting the spliced transcripts from the reference genome according to  the desired annotation (e.g. unfiltered, or filtered for certain classes of transcripts),as well as the collapsed intervals corresponding to the introns of genes. 
 
The intronic sequences of the *splici* reference play important roles in the various kinds of experiments discussed in the [manuscript](https://www.biorxiv.org/content/10.1101/2021.06.29.450377v2) of `alevin-fry`. For single-cell RNA-seq data, although one typically focuses on the  fragments and UMIs arising from the (spliced) transcriptome,and only considers the spliced and ambiguous counts when performing downstream analyses, the intronic sequences act similarly to decoy sequences proposed by [Srivastava 2020 et al.](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02151-8). They account for fragments that might otherwise selectively-align or pseudoalign to the transcriptome with lower quality, but that, in fact, derive from some unspliced RNA transcript molecule. *splici* improves the accuracy and false discovery of mapping, as the intronic sequences of *splici* act as the decoy sequence to absorb reads map to the spliced transcriptome with low quality.For single-nucleus RNA-seq and RNA velocity analysis, the usage is straightforward, as the intronic sequences enable `alevin-fry` to detect the signals from unspliced, immature RNA transcripts, which are crucial in those analyses.

To construct the *splici* reference, one provides a genome fasta file and a gene annotation GTF file of the  target species, as well as the read length of the experiments  to be processed. Here we describe the procedure to extract the  *splici* sequence of one gene; the procedure will be applied for  all genes specified in the provided GTF file.   First, the spliced transcript sequences are extracted for all isoforms of the gene. Next, the intronic regions of each isoform of  the gene are extracted.  
These intronic regions consist of the  full length of the unspliced transcript, subtracting out any  exonic sequence.  If alternative *splicing* positions occur within  exons, certain sub-intervals of a gene can appear as part of both  spliced transcripts and as intronic sequence. Additionally, each  extracted intronic interval is extended by the provided *flanking length* at its starting and ending position. By adding this *flanking length*, the reference can account for reads that map to the junction  of an exon and an intron.  Otherwise these reads cannot be confidently  mapped to either the unflanked intronic region or to the spliced  isoforms of the gene. 

The intronic regions of different isoforms often overlap considerably.   To avoid repetitive indexing of shared sub-intervals, the flanked  intronic regions across all isoforms of a gene will be combined if  they share overlapping sub-regions.  As the result, each unique  genomic sequence will appear only once in the combined intronic  regions of the gene.  The genomic sequences of those combined  intronic regions will then be extracted as the intronic part of this gene. These combined intronic sequences will each be given  a unique name, and all will be added to the *splici*  reference.

The process described above will be applied to all genes  defined in the provided GTF file.  These sequences will be  combined with any custom (user-provided) sequences, for example, one can include the sequence of  mitochondrial genes to avoid spurious mapping further. The resulting sequences constitute the *splici* reference. Note that sometimes one genomic region may belong to more than one gene. In this case, extracted sequences can appear multiple times in *splici* reference with different names. As duplicated sequences will anyway appear only once in the `salmon` index except if one specifically sets the `--keepDuplicates` flag (which we have not), we did not explicitly deduplicate repeated sequences across genes when constructing the *splici* reference.  However, this function accepts an optional argument, `dedup_seqs` that will perform this identical sequence deduplication during reference construction. 

Currently, the *splici* index depends on both the reference genome and annotation to be indexed, as well as the length of the reads that will be mapped against this index.  The read length is used to determine an appropriate flanking size (default of read length - 5) to add to the ends of the extracted collapsed intron sequences.  The read length used for construction need not exactly match those being mapped, and we have evaluated that the index is reasonably robust to similar read lengths and degrades gracefully as the indexed and mapped read lengths diverge.  Further, the short-read sequencing by synthesis most commonly employed for single-cell and single-nucleus experiments comprises a small set of characteristic read lengths.  Nonetheless, by simply constructing the index with a single flanking length that is as large as the longest reads to be considered, and by propagating the relative splice junction annotations to the mapping step, it should be possible for the index to be constructed independently of the read length parameter.  We are currently exploring the optimal implementation of this enhancement.






